### DISCUSSION TITLE
A second operation started on this context before a previous operation completed

### DISCUSSION THREAD

**Discussion by mdmoura** [quartznet/quartznet#964] [General] [answered]
I configured Quartz in an Asp.Net Core 3.1 project using:

        .ConfigureServices((context, services) => {

          services.AddDbContext<MyDbContext>(x => x.UseSqlServer(connectionString);

          services.AddQuartz(x => {

            x.SchedulerId = "Bot1";
            x.SchedulerName = "Bot";

            x.UseMicrosoftDependencyInjectionJobFactory(y => {
              y.AllowDefaultConstructor = true;
            });
              
            x.UseSimpleTypeLoader();
              
            x.UseInMemoryStore();
              
            x.UseDefaultThreadPool(y => {
              y.MaxConcurrency = 20;
            });

            x.AddJob<GetUserGenderJob>(y => y
              .StoreDurably()
              .WithIdentity("GetUserGender")
            );

            x.AddTrigger(t => t
              .WithIdentity("DailyAtTwoAm)   
                .ForJob("GetUserGender")
                // .WithCronSchedule("0 0 2 * * ? *")
                .WithCronSchedule("0/2 * * * * ? *")
            );

        });

        services.AddQuartzServer(y => {
          y.WaitForJobsToComplete = true;
        });

      });

And the job, just for testing, is:

```
  public class GetUserGenderJob : IJob {

    private readonly Context _context;

    public GetUserGenderJob(Context context) {
      _context = context;
    } 
    
    public async Task Execute(IJobExecutionContext context) {
      User user = await _context.Users.OrderBy(x => Guid.NewGuid()).FirstOrDefaultAsync();
      await Console.Out.WriteLineAsync(user.Name + " " + DateTime.UtcNow.ToString("HH:ss"));
    } 

  } 
```

I was able to run it without a problem a few times but in most of the runs I get the error:

```
System.InvalidOperationException: A second operation started on this context before a previous operation completed. 
This is usually caused by different threads using the same instance of DbContext.
```

It his bug or is something wrong with my configuration? 

I used the same DbContext configuration as on non-Quartz projects:

         services.AddDbContext<MyDbContext>(x => x.UseSqlServer(connectionString);

**Comment by lahma** [comment, marked as answer]
I think you might need to call `UseMicrosoftDependencyInjectionScopedJobFactory` instead of `UseMicrosoftDependencyInjectionJobFactory`. Quartz jobs are configured as transient and new instance is produced every time but the DbContext stuff might require the scoped job factory.

**Comment by mdmoura** [comment]
Yes, you are right DbContext is configured as Scoped. I didn't know about `UseMicrosoftDependencyInjectionScopedJobFactory` method. It is working now. Thank you for the help.

**Reply by lahma** [reply]
Great that you got it sorted, I updated [the documentation ](https://www.quartz-scheduler.net/documentation/quartz-3.x/packages/microsoft-di-integration.html#di-aware-job-factories)to note this more explicitly.
